File: algorithms/2_medium/S/00856_score_of_parentheses/stack.c
static inline int max_int(const int a, const int b)
{
    return a > b ? a : b;
}

int scoreOfParentheses(const char* s)
{
    assert(s != NULL && s[0] != '\0');
    const size_t s_len = strlen(s);
    const size_t stack_cap = s_len / 2;
    int* stack = malloc(stack_cap * sizeof(stack));
    assert(stack != NULL);
    size_t stack_len = 0;
    int score = 0;

    for (size_t i = 0; i < s_len; i += 1) {
        const char ch = s[i];

        if (ch == '(') {
            assert(stack_len < stack_cap && "invalid input");
            stack[stack_len] = 0;
            stack_len += 1;
        } else if (ch == ')') {
            assert(stack_len > 0 && "invalid input");
            const int curr = max_int(stack[stack_len - 1], 1);
            stack_len -= 1;
            if (stack_len == 0) {
                score += curr;
            } else {
                stack[stack_len - 1] += curr * 2;
            }
        } else {
            assert(false && "invalid input");
        }
    }

/* cleanup: */
    free(stack);
    return score;
}

File: algorithms/2_medium/S/00856_score_of_parentheses/stack.rs
impl Solution {
    pub fn score_of_parentheses(s: String) -> i32 {
        let mut stack: Vec<i32> = Vec::new();
        let mut score = 0_i32;

        for &byte in s.as_bytes() {
            match byte {
                b'(' => stack.push(0),
                b')' => {
                    let curr = 1_i32.max(stack.pop().expect("invalid input"));
                    if let Some(val) = stack.last_mut() {
                        *val += 2 * curr;
                    } else {
                        score += curr;
                    }
                }
                _ => unreachable!("invalid input"),
            }
        }

        score
    }
}
